tensorflow/compiler/xla/service/gpu/ir_emitter_nested.cc

```diff
@@ -278,12 +279,21 @@ llvm::Value* AddrCastToDefault(llvm::Value* arg, llvm::IRBuilder<>& b) {
   return arg;
 }
 
-void EmitAMDGPUAtomicAdd(llvm::IRBuilder<>* builder,
-                         llvm::Value* output_address, llvm::Value* source) {
+void EmitAMDGPUAtomicFAdd(llvm::IRBuilder<>* builder,
+                          IrEmitterContext& ir_emitter_context,
+                          llvm::Value* output_address, llvm::Value* source,
+                          PrimitiveType element_type) {
+  static bool paired_atomics = []() -> bool {
+    const char* flag = std::getenv("ROCM_PAIRED_ATOMICS");
+    return !(flag && (std::strcmp(flag, "false") == 0 ||
+                      std::strcmp(flag, "0") == 0));
+  }();
+
   CHECK(IsAMDGPU(builder->GetInsertBlock()->getModule()));
+  // For now we expect for BF16 op to be done trough FP32
```

Comment:
Why will bf16 op be done through fp32?

Reply:
XLA inesert conversion from to fp32 so the computation becomes more complex and doesn't take fast atomic path. Same on cuda. The CAS path when optimized shows similar performance to the intrinsic, so I chose to fix that.

Reply:
I am not sure if I understand the comment. Could you clarify further?

I am also seeing `CHECK_NE(element_type, BF16);`. Does it mean that you wanted to check that `element_type` is not equal to bf16?

Reply:
Yup. To crash if we reach this stage for bf16. Recuse I want to find out if it is happening in the wild. BF16 computation have extra conversions from to fp32 and fail here https://github.com/ROCm/tensorflow-upstream/blob/r2.14-rocm-enhanced/tensorflow/compiler/xla/service/gpu/ir_emitter_nested.cc#L383 if I remember correctly.